### ISSUE TITLE
Scanning inconsistently failing during attempt to descend into archives

### ISSUE THREAD

**Issue by mikjo** [dependency-check/DependencyCheck#570]
We are seeing occasional inconsistent errors while descending into archive files. This happens more frequently under load than otherwise. Running two or more scans in parallel on my system makes it likely that at least one of them will fail.

We are seeing this failure both within the gradle OWASP DC plugin and from the command line.

In a log from a failed gradle run I saw:

```
12:24:58.990 [DEBUG] [org.owasp.dependencycheck.Engine] Begin Analysis of '/tmp/dctemp/check3105223413902798852tmp/15/jacocoagent.jar'
12:24:59.009 [WARN] [org.owasp.dependencycheck.Engine] An error occurred while analyzing '/tmp/dctemp/check3105223413902798852tmp/15/jacocoagent.jar'.
12:24:59.011 [DEBUG] [org.owasp.dependencycheck.Engine] 
org.owasp.dependencycheck.analyzer.exception.AnalysisException: Exception occurred reading the JAR file.
        at org.owasp.dependencycheck.analyzer.JarAnalyzer.analyzeFileType(JarAnalyzer.java:246)
        at org.owasp.dependencycheck.analyzer.AbstractFileTypeAnalyzer.analyze(AbstractFileTypeAnalyzer.java:193)
        at org.owasp.dependencycheck.Engine.analyzeDependencies(Engine.java:401)
        at org.owasp.dependencycheck.Engine$analyzeDependencies$0.call(Unknown Source)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:117)
        at org.owasp.dependencycheck.gradle.tasks.Check.check(Check.groovy:79)
        ...
Caused by: java.io.FileNotFoundException: /tmp/dctemp/check3105223413902798852tmp/15/jacocoagent.jar (No such file or directory)
        at java.util.zip.ZipFile.open(Native Method)
        at java.util.zip.ZipFile.<init>(ZipFile.java:220)
        at java.util.zip.ZipFile.<init>(ZipFile.java:150)
        at java.util.jar.JarFile.<init>(JarFile.java:166)
        at java.util.jar.JarFile.<init>(JarFile.java:103)
        at org.owasp.dependencycheck.analyzer.JarAnalyzer.parseManifest(JarAnalyzer.java:657)
        at org.owasp.dependencycheck.analyzer.JarAnalyzer.analyzeFileType(JarAnalyzer.java:241)
        ... 133 more
```

I am running the following script (connection information obscured) to reproduce the error using the command-line scanner:

``` bash
#!/bin/bash
mkdir -p jar out
touch out/dependency-check-report.xml
curl -k http://central.maven.org/maven2/org/jacoco/org.jacoco.agent/0.7.1.201405082137/org.jacoco.agent-0.7.1.201405082137.jar > jar/org.jacoco.agent-0.7.1.201405082137.jar
while ! grep dctemp out/dependency-check-report.xml ; do
    rm  out/dependency-check-report.xml && ~/bin/dependency-check/bin/dependency-check.sh --noupdate --out out --scan jar -f ALL --project jacoco --log out/dc.log --connectionString jdbc:mysql://[...] --dbDriverName com.mysql.jdbc.Driver --dbDriverPath /usr/share/java/mysql-connector-java.jar --dbUser [...] --dbPassword [...] ;
    echo $? ;
    ls -l out/dependency-check-report.xml ;
done
```

~/bin/dependency-check/bin/dependency-check.sh is in an unaltered download of 1.4.3.

Along with many perfectly good runs, I occasionally get output like:

```
[INFO] Analysis Started
[WARN] An error occurred while analyzing '/.../org.jacoco.agent-0.7.1.201405082137.jar'.
[WARN] Unable to open jar file 'org.jacoco.agent-0.7.1.201405082137.jar: jacocoagent.jar'.
[WARN] An error occurred while analyzing '/tmp/dctemp/check4421204392009757457tmp/1/jacocoagent.jar'.
[INFO] Creating the CPE Index
[INFO] CPE Index Created (1572 ms)
[INFO] Analysis Complete (2656 ms)
[ERROR] Archive file was not found.
[ERROR] Exception occurred reading the JAR file.
242
-rw-rw-r--. 1 mikjo mikjo 10714 Sep 20 13:15 out/dependency-check-report.xml
```

A good run looks like:

```
[INFO] Analysis Started
[INFO] Creating the CPE Index
[INFO] CPE Index Created (6398 ms)
[INFO] Analysis Complete (8944 ms)
0
-rw-rw-r--. 1 mikjo mikjo 20348 Sep 20 13:15 out/dependency-check-report.xml
```

We have reproduced this with both HotSpot and OpenJDK.

I previously incorrectly reported this to be the same problem as #530 because of a similar error message, but the `grep dctemp` in the reproducer test case never finds any lines, so at the very least we are seeing a different symptom.  If my system is mostly idle, it rarely fails; if I run several scans in parallel it is likely to happen to at least one of the scans within a few minutes.

**Comment by mikjo** [comment]
Following suggestions from @jeremylong in #530 the author of our gradle plugin that invokes the `org.owasp:dependency-check-gradle` plugin reported to me that `failOnError=false` is set but it is still failing.

**Comment by eisakson** [comment]
I gave @mikjo bad info. We do not set failOnError to false but we do want to be able to turn down the failBuildOnCVSS value at some point. Checking my assumption:

failOnError = false will prevent this class of AnalysisException and other errors like a malformed suppression file from failing the build but would still allow a vulnerability to fail the build if we turned failBuildOnCVSS to zero.

Is my understanding correct?

**Comment by jeremylong** [comment]
@eisakson yes, your understanding is correct.

@mikjo I have tried a few times to reproduce the problem you have described. I've tested a script on Windows, Java 8 that opened 100 instances of dependency-check at the same time and I did not see the error you describe. I tried running your script on Ubuntu with openJDK 8 and it ran for 30 minutes with no error.  Now it could be that another change I made fixed this. Is there any chance you could compile and test the latest snapshot and rerun your test? If it fails, please let me know the exact OS and version of Java.

--Jeremy

**Comment by johnsonm** [comment]
I was @mikjo when I was at SAS. I have left SAS so I can't re-run the tests. However, I successfully reproduced the problem at least with OpenJDK on CentOS 7 and Fedora 23; possibly on Fedora 24 as well.

**Comment by wschiang** [comment]
@mikjo and I have been tracking the issues related to the temporary folder and the tmp/dctemp pattern in the filePath as part of the vulnerability reports.

We put org.jacoco.agent-0.7.1.201405082137.jar as downloaded from maven central in a directory, and started a loop running the cli scanner against that directory. We can always reproduce the issues after a couple of runs. We did this test with OWASP DC v1.4.3, and Java HotSpot “1.8.0_25”.

During the loop_test cycle, we observed three scenarios when job is ended: 
1)  run successfully, 
2)  run failed with return code 242, /tmp/dctemp directory disappeared during Jar Analyzer process, caused execution exception, 
3)  same as #2, but leaves /tmp/dctemp pattern in the filePath

**Comment by jeremylong** [comment]
Pretty sure I just figured out what is going on. The temporary directory was not unique to each execution of dependency-check and it tries to delete the temporary directory when done analyzing (thus another instance of dependency-check would have temp files in the folder and they would get deleted).  This has been fixed in 1.4.4-SNAPSHOT.

**Comment by bmekewar** [comment]
This issue is reappeared in 1.4.5

[DependencyCheck] Scanning: /disk1/jenkins/jobs/ACS-OASG-DATABASE-9.0-RPM/workspace
[DependencyCheck] Analyzing Dependencies
[DependencyCheck] One or more exceptions were thrown while executing Dependency-Check
[DependencyCheck] Exception Caught: org.owasp.dependencycheck.analyzer.exception.AnalysisException
[DependencyCheck] Cause: zip file is empty
[DependencyCheck] Message: Exception occurred reading the JAR file (oasg-12c-PSU-160419.zip: opatchauto-core.jar).
Build step 'Invoke OWASP Dependency-Check analysis' marked build as failure
[DependencyCheck] Skipping publisher since build result is FAILURE
[ANALYSIS-COLLECTOR] Skipping publisher since build result is FAILURE

Not sure how to fix it. I suspect if Zip file is big enough to scan (say 2GB) will temp folders still holds the extracted JAR reference folder for scan. 

But its mentioned its been fixed in 1.4.4 but re-appearing in 1.4.5.

**Comment by oosterholt** [comment]
This issue is reappeared in 3.1.2

What information do I need to add to make analyzing possible?
